test/integration/cluster_filter_integration_test.cc

```diff
@@ -27,7 +27,7 @@ class PoliteFilter : public Network::Filter, Logger::Loggable<Logger::Id::filter
   Network::FilterStatus onWrite(Buffer::Instance& data, bool end_stream) override {
     ENVOY_CONN_LOG(debug, "polite: onWrite {} bytes {} end_stream", write_callbacks_->connection(),
                    data.length(), end_stream);
-    if (!write_greeted_) {
+    if (!write_greeted_ && on_new_connection_called_ == 1) {
```

Comment:
sorry I'm a bit confused - is this new guard needed because on new connection called is now called more than once (and may be 2) or to only inject data if it's called (in which case I think we could replace with an assert)

Reply:
This change in the test is done solely to make it depend on the new `onNewConnection()` call behavior. Not sure an assert here is the right thing to do, as it could give an impression that `onWrite()` is never supposed to be called before `onNewConnection()`. To elaborate, `onNewConnection()` is a read filter callback and `onWrite()` can still be called before `onNewConnection()` if upstream data is sent before the connection is established.

Reply:
is onWrite allowed to be called before onNewConnection?  You could also add the assert and comment that this test mandates the ordering without saying the ordering is globally required.

Reply:
> is onWrite allowed to be called before onNewConnection? 

Yes. ConnectionImpl maintains a write buffer where all the data passed by the filter chain gets moved, and the case where the socket is still connecting is explicitly handled:
```
    // Activating a write event before the socket is connected has the side-effect of tricking
    // doWriteReady into thinking the socket is connected. On macOS, the underlying write may fail
    // with a connection error if a call to write(2) occurs before the connection is completed.
    if (!connecting_) {
      ioHandle().activateFileEvents(Event::FileReadyType::Write);
    }
```

Reply:
Reverted this particular change while adding runtime feature flag.